libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt

```diff
@@ -0,0 +1,158 @@
+        state = when {
+            hasRoutes && isDriving -> {
+                val state = ActiveGuidance
+                state.sessionId = TelemetryUtils.obtainUniversalUniqueIdentifier()
+                state
+            }
+            isDriving -> {
+                val state = FreeDrive
+                state.sessionId = TelemetryUtils.obtainUniversalUniqueIdentifier()
+                state
+            }
+            else -> {
+                val state = Idle
+                state.sessionId = ""
+                state
+            }
+        }
+    }
+
+    internal fun registerNavigationSessionStateObserver(
+        navigationSessionStateObserver: NavigationSessionStateObserver
+    ) {
+        stateObservers.add(navigationSessionStateObserver)
+        navigationSessionStateObserver.onNavigationSessionStateChanged(state)
+    }
+
+    internal fun unregisterNavigationSessionStateObserver(
+        navigationSessionStateObserver: NavigationSessionStateObserver
+    ) {
+        stateObservers.remove(navigationSessionStateObserver)
+    }
+
+    internal fun unregisterAllNavigationSessionStateObservers() {
+        stateObservers.clear()
+    }
+
+    override fun onRoutesChanged(routes: List<DirectionsRoute>) {
+        hasRoutes = routes.isNotEmpty()
+    }
+
+    override fun onSessionStateChanged(tripSessionState: TripSessionState) {
+        isDriving = when (tripSessionState) {
+            TripSessionState.STARTED -> true
+            TripSessionState.STOPPED -> false
+        }
+    }
+}
+
+/**
+ * Contains the various states that can occur during a navigation.
+ *
+ * The [MapboxNavigation] implementation can enter into the following session states:
+ * - [Idle]
+ * - [FreeDrive]
+ * - [ActiveGuidance]
+ *
+ * The SDK starts off in an [Idle] state.
+ * Whenever the [MapboxNavigation.startTripSession] is called, the SDK will enter the [FreeDrive] state.
+ * If the session is stopped, the SDK will enter the [Idle] state.
+ * If the SDK is in an [Idle] state, it stays in this same state even when a primary route is available.
+ * If the SDK is already in the [FreeDrive] mode or entering it whenever a primary route is available,
+ * the SDK will enter the [ActiveGuidance] mode instead.
+ * When the routes are manually cleared, the SDK automatically fall back to either [Idle] or [FreeDrive] state.
+ * When transitioning across states of a trip session the [sessionId] will change (empty when [Idle]).
+ */
+sealed class NavigationSessionState {
+
+    /**
+     * Random session UUID.
+     * This is generated internally based on the current state within an trip session.
+     * I.e. will change when transitioning across states of a trip session. Empty when [Idle].
+     *
+     * Useful to use it in combination with the [MapboxHistoryRecorder].
+     *
+     * @see [TripSessionState]
+     */
+    var sessionId: String = ""
+        internal set
+
+    /**
+     * Idle state
+     */
+    object Idle : NavigationSessionState() {
```

Comment:
Making this can object will overwrite IDs of past sessions, won't it?

If we get a state with an ID, and then the state goes through a cycle, the ID of our old cached value will be different.

This should rather be a class
```suggestion
    class Idle(sessionId: String) : NavigationSessionState(sessionId) {
```